File: Golf_Website/API/ScorecardController.cs
using Golf_Website.Models;
using Golf_Website.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Web.Http;

namespace Golf_Website.API
{
    public class ScorecardController : ApiController
    {
        private IGenericRepository _genRepo;

        public ScorecardController(IGenericRepository genRepo)
        {

            _genRepo = genRepo;

        }

        //Sending to api/scorecard
        public IHttpActionResult GetAllScorecards()
        {
            var data = _genRepo.Query<Scorecard>();

            return Ok(data);

        }

        [Route("api/scorecard/user/{userName}")]
        public IHttpActionResult GetCartItemsByUser(string userName)
        {
            var cartData = _genRepo.Query<Scorecard>().Where(sc => sc.UserName == userName);
            return Ok(cartData);
        }

        [Route("api/scorecard/course/{courseId}")]
        public IHttpActionResult GetCartItemsByCourse(int courseId)
        {
            var cartData = _genRepo.Query<Scorecard>().Where(sc => sc.CourseId == courseId).FirstOrDefault();
            return Ok(cartData);
        }

        public IHttpActionResult PostScorecard(Scorecard scorecardToAdd)
        {
            if (ModelState.IsValid)
            {

                if (scorecardToAdd.Id == 0)
                {
                    scorecardToAdd.ScoreDate = DateTime.Now;
                    _genRepo.Add<Scorecard>(scorecardToAdd);
                    _genRepo.SaveChanges();
                    return Ok();
                }
                else
                {
                    var cardToEdit = _genRepo.Find<Scorecard>(scorecardToAdd.Id);

                    cardToEdit.MyDifficulty = scorecardToAdd.MyDifficulty;
                    cardToEdit.MyScore1 = scorecardToAdd.MyScore1;
                    cardToEdit.MyScore2 = scorecardToAdd.MyScore2;
                    cardToEdit.MyScore3 = scorecardToAdd.MyScore3;
                    cardToEdit.MyScore4 = scorecardToAdd.MyScore4;
                    cardToEdit.MyScore5 = scorecardToAdd.MyScore5;
                    cardToEdit.MyScore6 = scorecardToAdd.MyScore6;
                    cardToEdit.MyScore7 = scorecardToAdd.MyScore7;
                    cardToEdit.MyScore8 = scorecardToAdd.MyScore8;
                    cardToEdit.MyScore9 = scorecardToAdd.MyScore9;
                    cardToEdit.MyScore10 = scorecardToAdd.MyScore10;
                    cardToEdit.MyScore11 = scorecardToAdd.MyScore11;
                    cardToEdit.MyScore12 = scorecardToAdd.MyScore12;
                    cardToEdit.MyScore13 = scorecardToAdd.MyScore13;
                    cardToEdit.MyScore14 = scorecardToAdd.MyScore14;
                    cardToEdit.MyScore15 = scorecardToAdd.MyScore15;
                    cardToEdit.MyScore16 = scorecardToAdd.MyScore16;
                    cardToEdit.MyScore17 = scorecardToAdd.MyScore17;
                    cardToEdit.MyScore18 = scorecardToAdd.MyScore18;
                    cardToEdit.MyFront = scorecardToAdd.MyScore1 + scorecardToAdd.MyScore2 + scorecardToAdd.MyScore3 + scorecardToAdd.MyScore4 + scorecardToAdd.MyScore5 + scorecardToAdd.MyScore6 + scorecardToAdd.MyScore7 + scorecardToAdd.MyScore8 + scorecardToAdd.MyScore9;
                    cardToEdit.MyBack = scorecardToAdd.MyScore10 + scorecardToAdd.MyScore11 + scorecardToAdd.MyScore12 + scorecardToAdd.MyScore13 + scorecardToAdd.MyScore14 + scorecardToAdd.MyScore15 + scorecardToAdd.MyScore16 + scorecardToAdd.MyScore17 + scorecardToAdd.MyScore18;
                    cardToEdit.MyTotal = cardToEdit.MyFront + cardToEdit.MyBack;
                    cardToEdit.HandicapAdj = scorecardToAdd.HandicapAdj;
                    _genRepo.SaveChanges();
                    return Ok();

                }

            }

            return BadRequest();
        }

        public IHttpActionResult DeleteScorecard(int id)
        {
            _genRepo.Delete<Scorecard>(id);
            _genRepo.SaveChanges();
            return Ok();
        }
    }
}

File: Golf_Website/ngApp/controllers/scorecardController.ts
namespace MyApp.Controllers {

    export class ScorecardController {

        public myScorecards;
        public userName;
        public woodforestScorecard;
        public scoreDateSelected;
        public items;
        public courseSelected;
        public selectedScorecard;
        public saveDisplay;
        public courseDetailsSelected;
        public showScore;
        public viewUser;
        public showAddScorecard;

        constructor(private scorecardService: MyApp.Services.ScorecardService, private accountService: MyApp.Services.AccountService, private $routeParams: ng.route.IRouteParamsService, private $location: ng.ILocationService, private $route: ng.route.IRouteService, private $uibModal: ng.ui.bootstrap.IModalService, private $filter: ng.IFilterDate, private courseService: MyApp.Services.CourseService) {
            this.saveDisplay = false;
            this.showScore = false;
            this.courseSelected = {
                id: null,
                difficulty: null,
                courseId: null
            };
            this.viewUser = $routeParams["userName"];
            this.scoreDateSelected = {
                courseId: 0
            };
            this.userName = accountService.getUserName();
            if (this.viewUser == this.userName) {
                this.showAddScorecard = true;
            } else {
                this.showAddScorecard = false;
            }
            this.myScorecards = this.scorecardService.listScorecardsByUserName(this.viewUser);

        }


        public showScores() {
            for (var i = 0; i < this.myScorecards.length; i++) {
                let id: number = (this.myScorecards[i].courseId);
                this.myScorecards[i].courseInfo = this.courseService.GetCourse(id);
            }
            

            this.showScore = true;
        }

        public selectScorecard(scorecard) {
            //Woodforest 4, Windrose 5, Carltonwoods 6
            return this.scorecardService.listScorecardsByCourse(scorecard);
        }

        public getCourseInfoById() {
            
            this.courseSelected.details = this.courseService.GetCourse(this.courseDetailsSelected.courseId);
        }

        public showSaveFeatures() {
            this.saveDisplay = true;
        }

        public updateScorecard(scorecardToSave) {
            scorecardToSave.myTotal = scorecardToSave.myScore1 + scorecardToSave.myScore2 + scorecardToSave.myScore3 + scorecardToSave.myScore4 + scorecardToSave.myScore5 + scorecardToSave.myScore6 + scorecardToSave.myScore7 + scorecardToSave.myScore8 + scorecardToSave.myScore9 + scorecardToSave.myScore10 + scorecardToSave.myScore11 + scorecardToSave.myScore12 + scorecardToSave.myScore13 + scorecardToSave.myScore14 + scorecardToSave.myScore15 + scorecardToSave.myScore16 + scorecardToSave.myScore17 + scorecardToSave.myScore18;
            if (scorecardToSave.myDifficulty == "Championship") {
                scorecardToSave.handicapAdj = (scorecardToSave.myTotal - scorecardToSave.courseInfo.champRating) * 113 / scorecardToSave.courseInfo.champSlope;
            } else if (scorecardToSave.myDifficulty == "Tournament") {
                scorecardToSave.handicapAdj = (scorecardToSave.myTotal - scorecardToSave.courseInfo.tournRating) * 113 / scorecardToSave.courseInfo.tournSlope;
            } else if (scorecardToSave.myDifficulty == "Members") {
                scorecardToSave.handicapAdj = (scorecardToSave.myTotal - scorecardToSave.courseInfo.membRating) * 113 / scorecardToSave.courseInfo.membSlope;
            } else if (scorecardToSave.myDifficulty == "Ladies") {
                scorecardToSave.handicapAdj = (scorecardToSave.myTotal - scorecardToSave.courseInfo.ladyRating) * 113 / scorecardToSave.courseInfo.ladySlope;
            }
            console.log(scorecardToSave);
            this.scorecardService.saveScorecard(scorecardToSave).then(() => {
                this.saveDisplay = false;
            });

        }

        public courseDetails() {
            //Not Working
            //console.log(this.scoreDateSelected.courseId);
            //this.courseDetailsSelected.$promise.then((result) => {
            //    console.log(result);
            //});

            //console.log(this.courseDetailsSelected);
            //console.log(this.courseDetailsSelected.courseId);
            //this.getCourseInfoById();
            //console.log(this.courseSelected.details);
        }

        public createScorecard() {
            
            this.selectedScorecard = this.selectScorecard(this.courseSelected.id).$promise.then((result) => {
                result.myDifficulty = this.courseSelected.difficulty;
                result.courseId = this.courseSelected.id;
                result.userName = this.userName;
                result.id = 0;
                result.myFront = 0;
                result.myBack = 0;
                result.myTotal = 0;
                this.scorecardService.saveScorecard(result).then(() => {
                    this.$route.reload();
                });

            });
            
            console.log(this.selectedScorecard);
        }

        public deleteScorecard(id) {
            this.scorecardService.deleteScorecard(id);
            this.$route.reload();
        }

        public addScoreDetailsModal(name) {
            var modalInstance = this.$uibModal.open({
                templateUrl: '/ngApp/views/modals/add-score-details.html',
                controller: AddScoreDetailsModal,
                controllerAs: "vm",
                resolve: {
                    name: () => { return name }
                },
                size: "sm"
            });
            
            modalInstance.result.then((course) => {
                this.courseSelected.id = course.id;
                this.courseSelected.difficulty = course.difficulty;
                this.createScorecard();
            });
        }

    }

    export class AddScoreDetailsModal {

        public passingData;
        public validationErrors;
        public course;

        constructor(private $uibModalInstance: ng.ui.bootstrap.IModalServiceInstance) {
            this.course = {
                id: null,
                difficulty: null,
                date: null
            }
            
        }

        public details() {
            if (!this.course.id || !this.course.difficulty) {
                this.validationErrors = "Please include all fields";
                return false;
            }
            this.$uibModalInstance.close(this.course);
        }

    }
}